beacon/beaclient/client.go

```diff
@@ -0,0 +1,141 @@
+
+package beaclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	eth2client "github.com/attestantio/go-eth2-client"
+	eth2api "github.com/attestantio/go-eth2-client/api"
+	eth2http "github.com/attestantio/go-eth2-client/http"
+	eth2spec "github.com/attestantio/go-eth2-client/spec"
+	"github.com/ethereum/go-ethereum/beacon/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/rs/zerolog"
+)
+
+var (
+	beaconLightClientBootstrap        = "eth/v1/beacon/light_client/bootstrap"
+	beaconLightClientUpdate           = "eth/v1/beacon/light_client/updates"
+	beaconLightClientOptimisticUpdate = "eth/v1/beacon/light_client/optimistic_update"
+	beaconLightClientFinalityUpdate   = "eth/v1/beacon/light_client/finality_update"
+)
+
+type Client struct {
+	ctx    context.Context
+	url    string
+	client eth2client.Service
+}
+
+func NewClient(ctx context.Context, server string) (*Client, error) {
+	client, err := eth2http.New(
+		ctx,
+		eth2http.WithAddress(server),
+		eth2http.WithLogLevel(zerolog.WarnLevel),
+		eth2http.WithEnforceJSON(true),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		ctx:    ctx,
+		url:    server,
+		client: client,
+	}, nil
+}
+
+func (c *Client) Bootstrap(root common.Hash) (*types.Bootstrap, error) {
+	resp, err := http.Get(fmt.Sprintf("%s/%s/%s", c.url, beaconLightClientBootstrap, root.String()))
+	if err != nil {
+		return nil, fmt.Errorf("failed http request: %w", err)
+	}
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+	var bs types.Bootstrap
+	if err := json.Unmarshal(b, &bs); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal data: %v", err)
+	}
+	return &bs, nil
+}
+
+func (c *Client) GetRangeUpdate(start, count int) ([]*types.LightClientUpdate, error) {
+	resp, err := http.Get(fmt.Sprintf("%s/%s?start_period=%d&count=%d", c.url, beaconLightClientUpdate, start, count))
+	if err != nil {
+		return nil, fmt.Errorf("failed http request: %w", err)
+	}
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+	var u []*types.LightClientUpdate
+	if err := json.Unmarshal(b, &u); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal data: %v", err)
+	}
+	return u, nil
+
+}
+
+func (c *Client) GetOptimisticUpdate() (*types.LightClientUpdate, error) {
+	resp, err := http.Get(fmt.Sprintf("%s/%s", c.url, beaconLightClientOptimisticUpdate))
+	if err != nil {
+		return nil, fmt.Errorf("failed http request: %w", err)
+	}
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %w", err)
+	}
+	var u types.LightClientUpdate
+	if err := json.Unmarshal(b, &u); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal data: %v", err)
+	}
+	return &u, nil
+}
+
+func (c *Client) GetFinalityUpdate() (*types.LightClientUpdate, error) {
```

Comment:
`GetRangeUpdate` `GetOptimisticUpdate` and `GetFinalityUpdate` kinda do the same thing, only for different URLs. We could factor that into a private method to reduce the amount of copy-pasta

Reply:
fixed!